tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,354 @@
+              autorestart feature implemented and values are the autorestart feature
+              state for that container
+    """
+    container_autorestart_states = {}
+
+    show_cmd_output = duthost.shell(CMD_CONTAINER_FEATURE_AUTORESTART)
+    for line in show_cmd_output["stdout_lines"]:
+        container_name = line.split()[0].strip()
+        container_state = line.split()[1].strip()
+        if container_state in ["enabled", "disabled"]:
+            container_autorestart_states[container_name] = container_state
+
+    return container_autorestart_states
+
+
+def get_disabled_container_list(duthost):
+    """
+    @summary: Get the container/service names which are disabled
+    @return: A list includes the names of disabled containers/services
+    """
+    disabled_containers = []
+
+    show_cmd_output = duthost.shell(CMD_FEATURE)
+    for line in show_cmd_output["stdout_lines"]:
+        container_name = line.split()[0].strip()
+        container_state = line.split()[1].strip()
+        if container_state == "disabled":
+            disabled_containers.append(container_name)
+
+    return disabled_containers
+
+
+def is_container_running(duthost, container_name):
+    """
+    @summary: Decide whether the container is running or not
+    @return:  Boolean value. True represents the container is running
+    """
+    is_running = duthost.shell("docker inspect -f \{{\{{.State.Running\}}\}} {}".format(container_name))
+    if is_running["stdout_lines"][0].strip() == "true":
+        return True
+    return False
+
+
+def check_container_state(duthost, container_name, should_be_running):
+    """
+    @summary: Determine whether a container is in the expected state (running/not running)
+    """
+    is_running = is_container_running(duthost, container_name)
+    if is_running == should_be_running:
+        return True
+    else:
+        return False
+
+
+def get_program_status(duthost, container_name, program_name):
+    """
+    @summary: Return the status of a program in the specified container
+    @return: "RUNNING" or "EXITED" represents the program is in the running
+             or exited status
+    """
+    program_status = None
+
+    process_list = duthost.shell("docker exec {} supervisorctl status".format(container_name))
+    for process_info in process_list["stdout_lines"]:
+        if process_info.find(program_name) != -1:
+            program_status = process_info.split()[1].strip()
+            break
+
+    return program_status
+
+
+def kill_process_by_pid(duthost, container_name, program_name, program_pid):
+    """
+    @summary: Kill a process in the specified container by its pid
+    """
+    duthost.shell("docker exec {} kill -SIGKILL {}".format(container_name, program_pid))
+
+    # Wait 30 seconds such that the container is stopped
+    wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+               CONTAINER_CHECK_INTERVAL_SECS,
+               check_container_state, duthost, container_name, False)
```

Comment:
In this function, the general idea is to kill the process by sending the `SIGKILL` signal and then double check whether it is actually killed or not. If the process is still running, the pytest script will show the message and exit. Usually when a critical process is killed, the corresponding container will be stopped within 3 or 4 seconds. 

If I did not use `wait_until(...)` at line 190, it is possible that the function at line 195 `is_container_running(...)` checked the container is still running. However, before the function `get_program_status(...)` is called, the container exited. In such scenario, the command `docker exec <container_name> supervisorctl status`  in the function `get_program_status(...)` will meet with an exception. So I used `wait_unti(...)` to expect that the container will exit after one critical process is killed before proceeding to check whether the critical process is still running.

If the container exited during the `wait_until(...)`, this is the expected behavior. If not, we will move forward to check the critical process is still running. If it is running, the pytest will fail and exit. If it is not running, another `wait_until` at the line 236 will continue to check whether the container will exit.

Reply:
But you are also calling this function in `verify_no_autorestart_with_non_critical_process()`, in which case you *don't* expect the container to stop. In that case, this function will always sleep CONTAINER_CHECK_INTERVAL_SECS, making the test runtime much longer.

Rather than ensuring the process was killed, can you simply check the return value from the `duthost.shell()` command to ensure the exit code is 0?

Reply:
Good suggestion! I will do that.

Reply:
Updated!